[core] add logic to request swap commitment

codingpeanut157 opened the pull request:
### **User description**
- SameMintSafeMode config var to configure whether or not wallet should compel the mint to commit before any Swap operation
- logic to request a commitment before any same-mint swap operation
- store_commitment in persistence


___

### **PR Type**
Enhancement


___

### **Description**
- Add `SameMintSafeMode` configuration to control swap commitment requests

- Implement `post_commitment` endpoint for requesting mint commitments before swaps

- Store commitment data in persistence layer (both in-memory and Rexie)

- Thread `SafeMode` parameter through wallet operations to enforce commitments

- Add error handling for Borsh serialization and signature verification


___

### Diagram Walkthrough


```mermaid
flowchart LR
  Config["SameMintSafeMode Config"]
  SafeMode["SafeMode Enum"]
  Commitment["post_commitment Endpoint"]
  Storage["Persistence Layer"]
  Wallet["Wallet Operations"]
  
  Config -->|configures| SafeMode
  SafeMode -->|passed to| Wallet
  Wallet -->|requests| Commitment
  Commitment -->|stores| Storage
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>config.rs</strong><dd><code>Add SameMintSafeMode configuration enum</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-71498585837598ae984b6f48bf3c8968bccc5d3c4c8163dddaeb83c5397e7d3e">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td><strong>lib.rs</strong><dd><code>Define TStamp type alias for timestamps</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-215a37c66b7040a141c225630bd02045caf1cf1f20ec88624fb33bbc5491f701">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mint.rs</strong><dd><code>Add post_commitment trait method to MintConnector</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-7eab6d3372222aa5e53932fef11d9369eb01d417f0536c5be1994dd751f47aad">+63/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>app.rs</strong><dd><code>Thread safe_mode through wallet initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-e720b0c2ecd74d4e7e02e6dededa253d5fad3026aacb07be25024bb8eeb4e586">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mod.rs</strong><dd><code>Define Commitment struct for persistence</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-18b4328796d3d2e0ab818c98bf5c4a17c7f39dccb7e27993745b28b9a0dd73b3">+10/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>inmemory.rs</strong><dd><code>Implement store_commitment for in-memory storage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-03626222bd1410ee277f551dd1992c24e781ffa7e3dc6db4adad9a4e5235a894">+22/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>rexie.rs</strong><dd><code>Add commitment store to Rexie database schema</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-de9d5fc4560252bae4acab8986721fd95d2a24ab38acf671f8024e3dc94e694c">+41/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mod.rs</strong><dd><code>Add SafeMode parameter to swap operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-75cb8e6a50ed45d750da38a0c0725794db0e7f0682a40c7a7446a0564b24fb5d">+46/-6</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>credit.rs</strong><dd><code>Thread SafeMode through credit pocket operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-9cfcfae3ef6fe248d343086e2dee504085dbc81457ec5e6b98a6a1ccc7f93616">+18/-5</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>debit.rs</strong><dd><code>Thread SafeMode through debit pocket operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-4e6f06900be2b435ce446adf18b2634ee588863817cfa96fd68972f5560659a5">+14/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>wallet.rs</strong><dd><code>Add SafeMode enum and thread through wallet methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-a728dbed0b385adf70434edb3f581a46a8ba286a76c7b6049469f1b36220caad">+79/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>utils.rs</strong><dd><code>Add compel_commitment utility function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-5aaa6d368f864dc60213651d4b19f97e68a5f4aa05517c87c2176d6350dbcf90">+41/-3</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Error handling</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>error.rs</strong><dd><code>Add Borsh and signature error handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-41a47c543cd4184039a7149e812dd0ce9d9d9ed9e9c1925516401b1b502769f8">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>rexie.rs</strong><dd><code>Add test for commitment storage in Rexie</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-e677dd1f824f16ea0c4fe7fddad041e2a830a03a886c42e472046e42dc158622">+20/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>Cargo.toml</strong><dd><code>Update bcr-common to master branch and add borsh</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Cargo.toml</strong><dd><code>Add bcr-common and borsh dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/Wallet-Core/pull/103/files#diff-663cb2a7378a6c3a4f5a754ee6cd1fd4f3e9add59e4080c64036662f5ee2e83d">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

zupzup commented on crates/bcr-wallet-core/src/utils.rs:
```diff
@@ -80,8 +81,33 @@ impl TryFrom<Proof> for ProofFingerprint {
+    match result {
+        Ok(response) => Ok(response),
+        // TODO: protest
+        // Err(Error::BorshSignature(BorshMsgSignatureError::Secp256k1(_))) => {},
```
why is this error commented out? leftover?

codingpeanut157 replied:
future implementation

we still need to design the protest protocol, yet I left this commented code as an indication of where "the plug" is

zupzup reviewed (approved):
LGTM 👍

codingpeanut157 committed:
[core] add logic to request swap commitment

- SameMintSafeMode config var to configure whether or not wallet should
  compel the mint to commit before any Swap operation
- logic to request a commitment before any same-mint swap operation
- store_commitment in persistence

stefanbitcr commented on Cargo.toml:
```diff
@@ -11,10 +11,11 @@ members = ["crates/bcr-wallet-core", "crates/bcr-wallet-lib"]
 anyhow = { version = "1" }
 async-trait = { version = "0.1" }
-bcr-common = {git = "https://github.com/BitcreditProtocol/bcr-common.git", tag = "v0.3.0"}
+bcr-common = {git = "https://github.com/BitcreditProtocol/bcr-common.git", branch = "master"}
```
For testing? Maybe release 0.6 ?

codingpeanut157 replied:
0.6 does not yet include some new features and APIs like `ProofFingerprint`

Once the project will settle a bit more I'll move to some more stable versioning system
